executor/write_test.go

```diff
@@ -1171,3 +1171,43 @@ func (s *testSuite) TestIssue4067(c *C) {
 	tk.MustExec("delete from t1 where id in (select id from t2)")
 	tk.MustQuery("select * from t1").Check(nil)
 }
+
+// Test issue https://github.com/pingcap/tidb/issues/4488
+func (s *testSuite) TestIssue4488(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1, t2, t3")
+
+	tk.MustExec("create table t1(a int, b int)")
+	tk.MustExec("insert into t1 set a=1, b=a+1")
+	tk.MustQuery("select a, b from t1").Check(testkit.Rows("1 2"))
+
+	tk.MustExec("create table t2(a int default 100, b int)")
+	tk.MustExec("insert into t2 set b=a+1, a=1")
+	tk.MustQuery("select a, b from t2").Check(testkit.Rows("1 101"))
+	tk.MustExec("insert into t2 (b) value (a)")
+	tk.MustQuery("select * from t2 where b = 100").Check(testkit.Rows("100 100"))
+
+	tk.MustExec("create table t3 (c int)")
+	tk.MustExec("insert into test.t3 set test.t3.c = '1'")
+	tk.MustQuery("select * from t3").Check(testkit.Rows("1"))
+}
+
+// Test issue https://github.com/pingcap/tidb/issues/4482
+func (s *testSuite) TestIssue4482(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t1")
+	tk.MustExec("create table t1 (a int not null)")
+	tk.MustExec("insert into t1 values (a+2)")
+	tk.MustExec("insert into t1 values (a)")
+	tk.MustQuery("select * from t1 order by a").Check(testkit.Rows("0", "2"))
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a bigint not null, b bigint not null)")
+	tk.MustExec("insert into t value(b + 1, a)")
+	tk.MustExec("insert into t set a = b + a, b = a + 1")
+	tk.MustExec("insert into t value(1000, a)")
+	tk.MustQuery("select * from t order by a").Check(testkit.Rows("0 1", "1 1", "1000 1000"))
```

Comment:
add the case issued by winoros in #4482 , `create table` without null flags.

Reply:
add test cases like:

```
mysql> create table t(a int default 1);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values (a);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)
```

Reply:
Please add the conner cases like `JSON`, `enum`, `key with auto increment` and so on.

Reply:
I have added these test cases at L1221-L1275.